platform/ios/src/MGLMapView.mm

```diff
@@ -4513,7 +4513,26 @@ - (void)notifyMapChange:(mbgl::MapChange)change
                 || self.userTrackingMode == MGLUserTrackingModeNone
                 || self.userTrackingState != MGLUserTrackingStateChanged)
             {
-                [self deselectAnnotation:self.selectedAnnotation animated:NO];
+                // Deselect annotation if it lies outside the viewport
```

Comment:
We already call `-visibleAnnotations` soon after this in order to reposition annotation views. We could refactor this code to call `-visibleAnnotations upfront and reuse the results for placement of annotation views, the callout view, and (on macOS) tooltip and cursor rects.